daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/ledgerinteraction/ide/UserManagementStore.scala

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf.engine.script.ledgerinteraction.ide
+
+import com.daml.lf.data.Ref.UserId
+import com.daml.lf.scenario.Error.UserManagementError
+import com.daml.ledger.api.domain.{User, UserRight}
+
+import scala.collection.mutable
+
+// Non thread safe, this is only used for execution of a sinlge script which is
+// single threaded.
+// Note (MK) This is a simplified version of the in-memory store used in the participant in
+// #11896. While it would be nice to not have to duplicate this for now this seems like the
+// simpler option than trying to reuse participant code in the script service.
+private[ledgerinteraction] class UserManagementStore {
+  import UserManagementError._
+  import UserManagementStore._
+
+  def createUser(user: User, rights: Set[UserRight]): Result[Unit] =
+    putIfAbsent(UserInfo(user, rights)) match {
+      case Some(_) => Left(UserExists(user.id))
+      case None => Right(())
+    }
+
+  def getUser(id: UserId): Result[User] =
+    lookup(id) match {
+      case Some(userInfo) => Right(userInfo.user)
+      case None => Left(UserNotFound(id))
+    }
+
+  def deleteUser(id: UserId): Result[Unit] =
+    dropExisting(id) match {
+      case Some(_) => Right(())
+      case None => Left(UserNotFound(id))
+    }
+
+  def grantRights(id: UserId, granted: Set[UserRight]): Result[Set[UserRight]] =
+    lookup(id) match {
+      case Some(userInfo) =>
+        val newlyGranted = granted.diff(userInfo.rights)
+        assert(replaceInfo(userInfo, userInfo.copy(rights = userInfo.rights ++ newlyGranted)))
```

Comment:
naive question. Why not pushing the assert inside `replaceInfo` ?

Reply:
No reason, I just copied that from the implementation in the participant. Agree that moving inside is better, done in 4f1b65f9d6901925164713267013ea2c4aac63bf